### PR TITLE
Private annotation

### LINKED ISSUE #175: Make the API closer to TF
We want to have an improved API for TFE. The goal is to be as close as possible to TF, so people don't have to learn another API on top of TF.

More precisely this issue focus on transforming the [TFE logistic regression example](https://github.com/mortendahl/tf-encrypted/blob/master/examples/logreg/tfe-logreg.py) so it looks the same as the [TF example](https://github.com/mortendahl/tf-encrypted/blob/master/examples/logreg/tf-logreg.py).

This include:
- [x] Add `reduce_sum` alias
- [x] Add `matmul` alias
- [x] Add a default LocalConfig that will be used by a protocol (Add a warning)
- [x] Add a new sessions system
- [x] Add a default protocol that will be used by default when calling operations (tfe.matmul ...)
- [x] Allows `define_private_input`to be used as an annotation

### PR DESCRIPTION
**Pull request by morgangiraud**
closes #175 
Look at the two files side by side `tf-logreg` and `tfe-logreg`.

Notice that now I print the accuracy "securely"

### CODE REVIEW

File: tensorflow_encrypted/protocol/pond.py [hunk 1]
```diff
@@ -187,21 +186,25 @@ def define_private_variable(
 
     def define_public_input(
         self,
-        provider: InputProvider,
+        player: Union[str, Player],
+        data_fn: Callable,
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/pond.py]: Think it's worth high lighting that it's a Callable that returns either a single tf.Tensor or list of tf.Tensors

File: tensorflow_encrypted/protocol/pond.py [hunk 2]
```diff
@@ -262,11 +269,15 @@ def helper(v: tf.Tensor):
 
     def define_output(
         self,
+        player: Union[str, Player],
         xs: Union['PondPrivateTensor', List['PondPrivateTensor']],
-        receiver: OutputReceiver,
+        post_fn: Callable,
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/pond.py]: could we call it `receiver_fn` instead? think that fits better with standard terminology

Comment (mortendahl) [reply tensorflow_encrypted/protocol/pond.py]: less important with a stronger type for the callable here, but could still be done

File: tensorflow_encrypted/protocol/pond.py [hunk 3]
```diff
@@ -218,14 +221,18 @@ def helper(v: tf.Tensor) -> 'PondPublicTensor':
 
     def define_private_input(
         self,
-        provider: InputProvider,
+        player: Union[str, Player],
+        data_fn: Callable,
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/pond.py]: could we call it `inputter_fn` instead? think that's closer to the standard mpc terminology of "input provider"

Comment (mortendahl) [reply tensorflow_encrypted/protocol/pond.py]: alternatively `provider_fn` but maybe that's less clear

Comment (mortendahl) [reply tensorflow_encrypted/protocol/pond.py]: stronger typing also important here (users might try to pass functions returning numpy arrays otherwise)

File: tensorflow_encrypted/protocol/pond.py (hunk 1 above)
Comment (mortendahl) [reply tensorflow_encrypted/protocol/pond.py]: see naming comments elsewhere, same applies here

File: examples/federated-average/run.py [hunk 4]
```diff
@@ -20,36 +20,34 @@
...
 server0 = config.get_player('server0')
 server1 = config.get_player('server1')
 crypto_producer = config.get_player('crypto_producer')
 
 with tfe.protocol.Pond(server0, server1, crypto_producer) as prot:
 
     # get input from inputters as private values
-    inputs = [prot.define_private_input(inputter)[0] for inputter in inputters]
+    inputs = [
+        prot.define_private_input(config.get_player(name), provide_input)[0] for name in player_names_inputter
```
Comment (mortendahl) [inline examples/federated-average/run.py]: you can avoid the `[0]` here by returning a `tf.Tensor` from `provide_input` instead of a list

File: examples/federated-average/run.py [hunk 5]
```diff
@@ -20,36 +20,34 @@
...
+    ]
 
     # sum all inputs and divide by count
     result = reduce(lambda x, y: x + y, inputs) * (1 / len(inputs))
 
     # send result to receiver
-    result_op = prot.define_output([result], result_receiver)
+    result_op = prot.define_output(config.get_player('result_receiver'), [result], receive_output)
```
Comment (mortendahl) [inline examples/federated-average/run.py]: is it possible to avoid call to `config.get_player` (and pass string instead)?

File: examples/mnist/run.py [hunk 6]
```diff
@@ -131,22 +137,22 @@ def receive_output(self, likelihoods: tf.Tensor) -> tf.Operation:
...
+    w0, b0, w1, b1 = params
     layer0 = prot.matmul(x, w0) + b0
     layer1 = prot.sigmoid(layer0 * 0.1)  # input normalized to avoid large values
-    layer2 = prot.matmul(layer1, w1) + b1
-    prediction = layer2
+    logits = prot.matmul(layer1, w1) + b1
 
     # send prediction output back to client
-    prediction_op = prot.define_output(prediction, prediction_client)
+    prediction_op = prot.define_output(prediction_client.player, [logits], prediction_client.receive_output)
```
Comment (mortendahl) [inline examples/mnist/run.py]: should be `logits` instead of `[logits]`, no?

File: tensorflow_encrypted/protocol/pond.py [hunk 7]
```diff
@@ -262,11 +273,16 @@ def helper(v: tf.Tensor):
 
     def define_output(
         self,
+        player: Union[str, Player],
         xs: Union['PondPrivateTensor', List['PondPrivateTensor']],
-        receiver: OutputReceiver,
+        post_fn: Callable[..., Any],
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/pond.py]: this guy is a `Callable[tf.Tensor, ...]`if i understand that syntax correct

Comment (morgangiraud) [reply tensorflow_encrypted/protocol/pond.py]: Not really, it can take any number of parameters:

`def post_fn(*args: tf.Tensor) -> Any`

But I didn't find any way to type this `*args` in `Callable` definition. I would need something like `Callable[...:tf.Tensor, Any]` or `Callable[StarArg: tf.Tensor, Any]` but it's not working

Also, I just noticed I didn't rename it, what name would you prefer? `outputter_fn`?

File: examples/mnist/run.py (hunk 6 above)
Comment (morgangiraud) [reply examples/mnist/run.py]: Both are allowed. I can change it if you prefer?

File: examples/federated-average/run.py (hunk 5 above)
Comment (morgangiraud) [reply examples/federated-average/run.py]: Yes

File: tensorflow_encrypted/protocol/pond.py (hunk 7 above)
Comment (mortendahl) [reply tensorflow_encrypted/protocol/pond.py]: yeah, `outputter_fn` works 👍
